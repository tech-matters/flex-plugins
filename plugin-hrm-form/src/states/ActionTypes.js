export const HANDLE_BLUR = 'HANDLE_BLUR';
export const HANDLE_FOCUS = 'HANDLE_FOCUS';
export const HANDLE_CHANGE = 'HANDLE_CHANGE';
export const INITIALIZE_CONTACT_STATE = 'INITIALIZE_CONTACT_STATE';
export const REMOVE_CONTACT_STATE = 'REMOVE_CONTACT_STATE';
export const SAVE_END_MILLIS = 'SAVE_END_MILLIS';
export const SAVE_CONTACT_STATE = 'SAVE_CONTACT_STATE';
