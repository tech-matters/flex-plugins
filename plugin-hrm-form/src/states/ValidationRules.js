import cloneDeep from 'lodash/cloneDeep';

import callTypes from './DomainConstants';
import { FieldType, ValidationType } from './ContactFormStateFactory';
import { getConfig } from '../HrmFormPlugin';

/*
 * Questionable whether we should export this
 * Be careful, this returns true if it's empty
 */
export function isNonDataCallType(callType) {
  return callType !== callTypes.caller && callType !== callTypes.child;
}

export function isNotCategory(value) {
  const notCategory = ['error', 'touched', 'type', 'validation', 'color'];
  return notCategory.includes(value);
}

export function isNotSubcategory(value) {
  const notSubcategory = ['type', 'color'];
  return notSubcategory.includes(value);
}

export function countSelectedCategories(categoryFormSection) {
  let count = 0;
  for (const category of Object.keys(categoryFormSection).filter(key => !isNotCategory(key))) {
    for (const subcategory of Object.keys(categoryFormSection[category]).filter(key => !isNotSubcategory(key))) {
      if (categoryFormSection[category][subcategory].value) {
        count += 1;
      }
    }
  }
  return count;
}

// NOTE: MODIFIES INPUT
function handleCallerOrChildInformationKeys(formToModify, ignoreTouched) {
  const { strings } = getConfig();

  Object.keys(formToModify)
    .filter(key => key !== 'type' && key !== 'validation' && key !== 'error')
    .forEach(key => {
      if (formToModify[key].type === FieldType.INTERMEDIATE) {
        handleCallerOrChildInformationKeys(formToModify[key], ignoreTouched);
      } else if (formToModify[key].validation) {
        if (formToModify[key].validation.includes(ValidationType.REQUIRED)) {
          const field = formToModify[key];
          if (field.type === FieldType.CHECKBOX_FIELD) {
            if ((ignoreTouched || field.touched) && countSelectedCategories(field) === 0) {
              field.error = strings['Error-CategoryRequired'];
              field.touched = true;
            } else {
              field.error = null;
            }
          } else if ((ignoreTouched || field.touched) && !field.value) {
            // explicitly set it to touched so it can't get unset later
            field.touched = true;
            field.error = 'This field is required';
          } else {
            field.error = null;
          }
        }
      }
    });
}

function validateCallerInformation(original, callType, ignoreTouched) {
  if (callType !== callTypes.caller) {
    return original;
  }
  const newForm = cloneDeep(original);
  handleCallerOrChildInformationKeys(newForm, ignoreTouched);
  return newForm;
}

function validateChildInformation(original, callType, ignoreTouched) {
  const newForm = cloneDeep(original);
  handleCallerOrChildInformationKeys(newForm, ignoreTouched);
  return newForm;
}

function validateCaseInformation(original, callType, ignoreTouched) {
  const newForm = cloneDeep(original);
  handleCallerOrChildInformationKeys(newForm, ignoreTouched);
  return newForm;
}

function validate(form, ignoreTouched = false) {
  if (isNonDataCallType(form.callType.value)) {
    return form;
  }
  let newForm = {};
  newForm = {
    ...form,
    callerInformation: validateCallerInformation(form.callerInformation, form.callType.value, ignoreTouched),
    childInformation: validateChildInformation(form.childInformation, form.callType.value, ignoreTouched),
    caseInformation: validateCaseInformation(form.caseInformation, form.callType.value, ignoreTouched),
  };
  return newForm;
}

export function validateBeforeSubmit(form) {
  if (isNonDataCallType(form.callType.value)) {
    return form;
  }

  /*
   * we need to also consider whether we care about caller information or just child
   * and if you start with caller type, get invalid, then switch to child type, it's still disabled
   */
  return validate(form, true);
}

/*
 * returns a new version of the form that updates the error messages
 * should this actually check for whether anything has changed?
 * not planning to do that yet, but could add it as an optimization
 */
export function validateOnBlur(form) {
  return validate(form);
}

// walk a form tree looking for non-null error values
export function formIsValid(form) {
  if ('callType' in form) {
    if (isNonDataCallType(form.callType.value)) {
      return true;
    }
  }

  if ('error' in form && form.error !== null) {
    return false;
  }
  for (const key of Object.keys(form)) {
    if (typeof form[key] === 'object' && form[key] !== null && !formIsValid(form[key])) {
      return false;
    }
  }
  return true;
}

export function moreThanThreeCategoriesSelected(categoryFormSection) {
  return countSelectedCategories(categoryFormSection) > 3;
}
