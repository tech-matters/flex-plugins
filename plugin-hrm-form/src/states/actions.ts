import {
  GeneralActionType,
  INITIALIZE_CONTACT_STATE,
  RECREATE_CONTACT_STATE,
  REMOVE_CONTACT_STATE,
  ContactFormDefinition,
} from './types';

export const initializeContactState = (definitions: ContactFormDefinition) => (taskId: string): GeneralActionType => ({
  type: INITIALIZE_CONTACT_STATE,
  definitions,
  taskId,
});

export const recreateContactState = (definitions: ContactFormDefinition) => (taskId: string): GeneralActionType => ({
  type: RECREATE_CONTACT_STATE,
  definitions,
  taskId,
});

export const removeContactState = (taskId: string): GeneralActionType => ({
  type: REMOVE_CONTACT_STATE,
  taskId,
});
