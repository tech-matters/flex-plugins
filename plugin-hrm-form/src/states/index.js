import { combineReducers } from 'redux';

import { reduce as ContactStateReducer } from './ContactState';
import { reduce as SearchFormReducer } from './SearchContact';
import { reduce as ConnectedCaseReducer } from './case/reducer';
import { reduce as QueuesStatusReducer } from './queuesStatus/reducer';
import { reduce as ConfigurationReducer } from './configuration/reducer';
import { reduce as RoutingReducer } from './routing/reducer';

// Register your redux store under a unique namespace
export const namespace = 'plugin-hrm-form';
export const contactFormsBase = 'activeContacts';
export const searchContactsBase = 'searchContacts';
export const connectedCaseBase = 'connectedCase';
export const queuesStatusBase = 'queuesStatusState';
export const configurationBase = 'configuration';
export const routingBase = 'routing';

// Combine the reducers
export default combineReducers({
  [contactFormsBase]: ContactStateReducer,
  [searchContactsBase]: SearchFormReducer,
  [queuesStatusBase]: QueuesStatusReducer,
  [connectedCaseBase]: ConnectedCaseReducer,
  [configurationBase]: ConfigurationReducer,
  [routingBase]: RoutingReducer,
});
