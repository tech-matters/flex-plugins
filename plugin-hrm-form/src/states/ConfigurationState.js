import { CHANGE_LANGUAGE } from './ActionTypes';

// default language to initialize plugin
export const defaultLanguage = 'en-US';

export const changeLanguage = language => ({ type: CHANGE_LANGUAGE, language });

const initialState = {
  language: defaultLanguage,
};

export function reduce(state = initialState, action) {
  switch (action.type) {
    case CHANGE_LANGUAGE:
      return {
        ...state,
        language: action.language,
      };
    default:
      return state;
  }
}
