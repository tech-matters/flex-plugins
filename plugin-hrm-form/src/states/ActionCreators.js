import { HANDLE_BLUR, HANDLE_FOCUS, SAVE_CONTACT_STATE } from './ActionTypes';
import { formIsValid, moreThanThreeCategoriesSelected, validateBeforeSubmit, validateOnBlur } from './ValidationRules';

export const handleBlur = dispatch => (form, taskId) => () => {
  console.log('Received blur event');
  dispatch({
    type: HANDLE_BLUR,
    form: validateOnBlur(form),
    taskId,
  });
};

export const handleFocus = dispatch => (taskId, parents, name) => {
  console.log(`handleFocus called with taskId = ${taskId}, parents = ${JSON.stringify(parents)}, name = ${name}`);
  dispatch({
    type: HANDLE_FOCUS,
    parents,
    name,
    taskId,
  });
};

export const handleSubmit = dispatch => (form, hrmBaseUrl, workerSid, helpline, handleCompleteTask) => task => {
  const newForm = validateBeforeSubmit(form);
  dispatch({
    type: HANDLE_BLUR, // probably need to rename this
    form: newForm,
    taskId: task.taskSid,
  });
  if (formIsValid(newForm)) {
    dispatch({
      type: SAVE_CONTACT_STATE,
      hrmBaseUrl,
      task,
      abortFunction: () => null,
      workerSid,
      helpline,
    });
    handleCompleteTask(task.taskSid, task);
  } else {
    window.alert('There is a problem with your submission.  Please check the form for errors.');
  }
};

/*
 * This is not technically an ActionCreator.  It's more of a filter and/or validator,
 * but it happens on an event.  Does it belong here?
 * This is also (dangerously?) hardcoding where the category information is.
 */
export const handleCategoryToggle = (form, handleChange) => (taskId, category, subcategory, newValue) => {
  const candidateCategories = {
    ...form.caseInformation.categories,
    [category]: {
      ...form.caseInformation.categories[category],
      [subcategory]: {
        ...form.caseInformation.categories[category][subcategory],
        value: newValue,
      },
    },
  };
  if (moreThanThreeCategoriesSelected(candidateCategories)) {
    window.alert('You cannot select more than three category options');
  } else {
    handleChange(taskId, ['caseInformation', 'categories', category], subcategory, newValue);
  }
};
