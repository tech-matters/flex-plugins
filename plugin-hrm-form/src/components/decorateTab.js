import React from 'react';
import ErrorIcon from '@material-ui/icons/Error';

import { StyledTab } from '../Styles/HrmStyles';
import { formIsValid } from '../states/ValidationRules';

const decorateTab = (label, formRoot) => {
  if (formIsValid(formRoot)) {
    return <StyledTab key={label} label={label} />;
  }
  return (
    <StyledTab
      key={label}
      label={
        <>
          <span style={{ verticalAlign: 'middle' }}>
            <ErrorIcon fontSize="inherit" style={{ color: 'red' }} />
          </span>{' '}
          {label}{' '}
        </>
      }
    />
  );
};

decorateTab.displayName = 'decorateTab';

export default decorateTab;
