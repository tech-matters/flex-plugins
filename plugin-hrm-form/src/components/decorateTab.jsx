import React from 'react';
import ErrorIcon from '@material-ui/icons/Error';

import { StyledTab } from '../styles/HrmStyles';
import { formIsValid } from '../states/ValidationRules';

const decorateTab = (label, formRoot) => {
  if (formIsValid(formRoot)) {
    return <StyledTab key={label} label={label} />;
  }
  return (
    <StyledTab
      key={label}
      label={
        <>
          <div style={{ verticalAlign: 'middle' }}>
            <ErrorIcon fontSize="inherit" style={{ color: 'red' }} />
          </div>{' '}
          {label}{' '}
        </>
      }
    />
  );
};

decorateTab.displayName = 'decorateTab';

export default decorateTab;
