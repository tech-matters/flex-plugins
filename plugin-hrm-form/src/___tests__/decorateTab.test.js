import renderer from 'react-test-renderer';

import { formIsValid } from '../states/ValidationRules';
import decorateTab from '../components/decorateTab';

jest.mock('../styles/HrmStyles', () => ({ StyledTab: 'StyledTab' }));
jest.mock('../states/ValidationRules', () => ({ formIsValid: jest.fn() }));
jest.mock('@material-ui/icons/Error', () => 'ErrorIcon');

test('decorateTab when valid', () => {
  formIsValid.mockReturnValueOnce(true);
  const component = renderer.create(decorateTab('My Label', {}));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('decorateTab when invalid', () => {
  formIsValid.mockReturnValueOnce(false);
  const component = renderer.create(decorateTab('My Label', {}));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
