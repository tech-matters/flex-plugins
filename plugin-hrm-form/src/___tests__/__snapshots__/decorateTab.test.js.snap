// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorateTab when invalid 1`] = `
<StyledTab
  label={
    <React.Fragment>
      <div
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        <ErrorIcon
          fontSize="inherit"
          style={
            Object {
              "color": "red",
            }
          }
        />
      </div>
       
      My Label
       
    </React.Fragment>
  }
/>
`;

exports[`decorateTab when valid 1`] = `
<StyledTab
  label="My Label"
/>
`;
